### PR TITLE
Add File::can_read_line test

### PR DESCRIPTION
**Pull request by Coderdreams**
A small test for a bug that was recently fixed in #5907

### CODE REVIEW

**Comment by alimpfard** [comment]
Looks like it's failing when run inside serenity?
```
 ==== Running test 33 out of 112 -- /usr/Tests/LibCore/TestLibCoreIODevice ====
/usr/Tests/LibCore/TestLibCoreIODevice
  Running test 'file_readline'.
  Failed to open /res/txt/long-line.txt: No such file or directory
  FAIL: ../Tests/LibCore/TestLibCoreIODevice.cpp:17: VERIFY_NOT_REACHED() called
  Shell: Job 1 (/usr/Tests/LibCore/TestLibCoreIODevice) Aborted
  
==== Added /usr/Tests/LibCore/TestLibCoreIODevice to list of failed tests. Failed tests so far: /usr/Tests/LibCore/TestLibCoreIODevice ====
Error:  /usr/Tests/LibCore/TestLibCoreIODevice returned non-zero exit code, check logs!
```

File: Tests/LibCore/TestLibCoreIODevice.cpp [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <LibCore/File.h>
+#include <LibTest/TestCase.h>
+#include <unistd.h>
+
+TEST_CASE(file_readline)
+{
+    auto path = "/res/txt/long-line.txt";
```
Comment (alimpfard) [inline Tests/LibCore/TestLibCoreIODevice.cpp:13]: This file is called `long-lines.txt`, not `long-line.txt`.

Comment (alimpfard) [reply Tests/LibCore/TestLibCoreIODevice.cpp:13]: Also, this seems to be written to be run only inside serenity, is it not being run with Lagom?

Comment (Coderdreams) [reply Tests/LibCore/TestLibCoreIODevice.cpp]: > This file is called `long-lines.txt`, not `long-line.txt`.

Fixed

Comment (Coderdreams) [reply Tests/LibCore/TestLibCoreIODevice.cpp]: > Also, this seems to be written to be run only inside serenity, is it not being run with Lagom?

Working on it, I'm getting:

/usr/bin/ld: Userland/Libraries/LibCore/libcore.so: error adding symbols: file in wrong format

**Review by linusg — COMMENTED** [review]
I don't think the test file should be in `Base/res`, `
Tests/LibCore` and installing via CMake seems more appropriate (we already do this for `test.frm`) - something like `install(FILES long_lines.txt DESTINATION usr/Tests/LibCore)`.

**Comment by pbaena-optimous** [comment]
> I don't think the test file should be in `Base/res`, ` Tests/LibCore` and installing via CMake seems more appropriate (we already do this for `test.frm`) - something like `install(FILES long_lines.txt DESTINATION usr/Tests/LibCore)`.

Fixed.

File: Tests/LibCore/TestLibCoreIODevice.cpp (hunk 1 above)
Comment (pbaena-optimous) [reply Tests/LibCore/TestLibCoreIODevice.cpp]: > Also, this seems to be written to be run only inside serenity, is it not being run with Lagom?

I think I managed to make it run with Lagom:

      Start  4: test-iodevice
 4/89 Test  #4: test-iodevice ........................   Passed    0.00 sec

**Comment by Coderdreams** [comment]
Please review, all suggestions were implemented as requested and tests run. Thanks in advance.

File: Base/home/anon/tests/run-tests-and-shutdown.sh [hunk 2]
```diff
@@ -25,10 +31,10 @@ run(index) {
 #     UserEmulator: Tests designed to run inside the Userspace Emulator
 #     stack-smash: Intentionally crashes by smashing the stack
 #     TestJSON: AK/TestJSON makes assumptions about $PWD to load its input files
-#     .frm: Test inputs that are not tests
+#     .frm,.txt: Test inputs that are not tests
 #     test-web: Requires the window server in order to work
 #     test-js: We start this one manually with the show progress flag set to false
-exclude_patterns='Kernel/Legacy|.inc|UserEmulator|stack-smash|TestJSON|.frm|test-web|test-js'
+exclude_patterns='Kernel/Legacy|.inc|UserEmulator|stack-smash|TestJSON|.txt|.frm|test-web|test-js'
```
Comment (linusg) [inline Base/home/anon/tests/run-tests-and-shutdown.sh:37]: now that the current working directory is set correctly, can we remove TestJSON from the exclude patterns?

File: Meta/Lagom/CMakeLists.txt [hunk 3]
```diff
@@ -191,6 +191,17 @@ if (BUILD_LAGOM)
...
+        set_target_properties(test-iodevice PROPERTIES OUTPUT_NAME test-iodevice)
+        target_link_libraries(test-iodevice Lagom)
+        target_link_libraries(test-iodevice LagomTest)
+        target_link_libraries(test-iodevice stdc++)
+        add_test(
+            NAME test-iodevice
+            COMMAND test-iodevice
+            WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/../../Tests/LibCore
+        )
+
```
Comment (linusg) [inline Meta/Lagom/CMakeLists.txt:204]: I don't think we need this in Lagom. Does it even work? The `add_executable()` path seems to be missing a `.cpp` suffix.

File: Userland/Shell/Tests/heredocs.sh [hunk 4]
```diff
@@ -1,6 +1,8 @@
 #!/bin/sh
 
 source $(dirname "$0")/test-commons.inc
+# go to a writable directory
+cd /tmp
```
Comment (linusg) [inline Userland/Shell/Tests/heredocs.sh:5]: Please:

- Put this in a separate commit
- Explain in the commit message why this is necessary

Comment (Coderdreams) [reply Userland/Shell/Tests/heredocs.sh:5]: Are you sure the commit linter isn't going to complain about having a separate commit? It was the reason I squashed in the first place.

Comment (linusg) [reply Userland/Shell/Tests/heredocs.sh:5]: No, as long as all commits follow the commit message rules, there won't be a complaint. Plenty of PRs have multiple commits :)

File: Meta/Lagom/CMakeLists.txt (hunk 3 above)
Comment (Coderdreams) [reply Meta/Lagom/CMakeLists.txt:204]: This works, if for example I change the filename I get a failure:
      Start  4: test-iodevice
 4/89 Test  #4: test-iodevice ........................Child aborted***Exception:   0.17 sec

It doesn't need the cpp extension. 

Also I was following the suggestion of alimpfard to add it to Lagom.

Comment (linusg) [reply Meta/Lagom/CMakeLists.txt:204]: Fine, but please add the cpp extension, like we do for literally every other `add_executable()` here as well.

File: Base/home/anon/tests/run-tests-and-shutdown.sh (hunk 2 above)
Comment (Coderdreams) [reply Base/home/anon/tests/run-tests-and-shutdown.sh:37]: It's out of scope and it's taking already too long to merge this PR, can we do it in another PR?

Comment (linusg) [reply Base/home/anon/tests/run-tests-and-shutdown.sh:37]: Sure.

File: Meta/Lagom/CMakeLists.txt (hunk 3 above)
Comment (Coderdreams) [reply Meta/Lagom/CMakeLists.txt]: Done

File: Userland/Shell/Tests/heredocs.sh (hunk 4 above)
Comment (Coderdreams) [reply Userland/Shell/Tests/heredocs.sh:5]: Ah OK, I'm still getting used to serenity commit policies, it seemed to indicate to squash everything after reviews.

**Comment by Coderdreams** [comment]
Everything is according to suggestions now, can we please merge this PR?

**Comment by Coderdreams** [comment]
Is there no interest in this test? I did almost everything that was requested already a long time ago.

**Comment by alimpfard** [comment]
Sorry I forgot about this PR ^^'
Could you rebase on top of master so we could make sure the CI still passes?
Also, could you please fixup the commit messages and make the first character uppercase (like "LibCore: Add unit test for File::read_line") while you're at it?

**Comment by Coderdreams** [comment]
Done fixing the commit messages and merging upstream master.

**Comment by Coderdreams** [comment]
I fixed the tests that fail because they can't write to the execution directory.

**Comment by alimpfard** [comment]
Please rebase instead of merging (the merge commit has to go)

**Comment by Coderdreams** [comment]
> Please rebase instead of merging (the merge commit has to go)

When I do a rebase of the merge commit a ton of commits appear on my PR, is that OK? Or do I need to delete the merge commit?

**Comment by Coderdreams** [comment]
> Please rebase instead of merging (the merge commit has to go)

Done, thanks.

**Comment by Coderdreams** [comment]
All tests pass again! \o/
